contracts/badger-ren/BadgerRenAdapter.sol

```diff
@@ -0,0 +1,272 @@
+import "deps/@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "deps/@openzeppelin/contracts-upgradeable/math/SafeMathUpgradeable.sol";
+import "deps/@openzeppelin/contracts-upgradeable/utils/ReentrancyGuardUpgradeable.sol";
+
+interface IGateway {
+    function mint(
+        bytes32 _pHash,
+        uint256 _amount,
+        bytes32 _nHash,
+        bytes calldata _sig
+    ) external returns (uint256);
+
+    function burn(bytes calldata _to, uint256 _amount) external returns (uint256);
+}
+
+interface IGatewayRegistry {
+    function getGatewayBySymbol(string calldata _tokenSymbol) external view returns (IGateway);
+
+    function getGatewayByToken(address _tokenAddress) external view returns (IGateway);
+
+    function getTokenBySymbol(string calldata _tokenSymbol) external view returns (IERC20);
+}
+
+interface ICurveExchange {
+    function exchange(
+        int128 i,
+        int128 j,
+        uint256 dx,
+        uint256 min_dy
+    ) external;
+
+    function get_dy(
+        int128,
+        int128 j,
+        uint256 dx
+    ) external view returns (uint256);
+
+    function calc_token_amount(uint256[2] calldata amounts, bool deposit) external returns (uint256 amount);
+
+    function add_liquidity(uint256[2] calldata amounts, uint256 min_mint_amount) external;
+
+    function remove_liquidity(uint256 _amount, uint256[2] calldata min_amounts) external;
+
+    function remove_liquidity_imbalance(uint256[2] calldata amounts, uint256 max_burn_amount) external;
+
+    function remove_liquidity_one_coin(
+        uint256 _token_amounts,
+        int128 i,
+        uint256 min_amount
+    ) external;
+}
+
+contract BadgerRenAdapter is OwnableUpgradeable, ReentrancyGuardUpgradeable {
+    using SafeMathUpgradeable for uint256;
+    using SafeERC20 for IERC20;
+
+    IERC20 renBTC;
+    IERC20 wBTC;
+
+    // RenVM gateway registry.
+    IGatewayRegistry public registry;
+    // Curve exchange contract for the renBTC/wBTC pool.
+    ICurveExchange public exchange;
+
+    event RecoverStuckRenBTC(uint256 amount, uint256 fee);
+    event MintRenBTC(uint256 amount, uint256 fee);
+    event BurnRenBTC(uint256 amount, uint256 fee);
+    event MintWBTC(uint256 renbtc_minted, uint256 wbtc_exchanged, uint256 fee);
+    event BurnWBTC(uint256 wbtc_transferred, uint256 renbtc_burned, uint256 fee);
+    event ExchangeWBTCBytesError(bytes error);
+    event ExchangeWBTCStringError(string error);
+
+    address public rewards;
+    address public governance;
+
+    uint256 public mintFeeBps;
+    uint256 public burnFeeBps;
+    uint256 private percentageFeeRewardsBps;
+    uint256 private percentageFeeGovernanceBps;
+
+    uint256 public constant MAX_BPS = 10000;
+
+    function initialize(
+        address _governance,
+        address _rewards,
+        address _registry,
+        address _exchange,
+        address _wbtc,
+        uint256[4] memory _feeConfig
+    ) public {
```

Comment:
Minor risk: suggestion to check the values on ctor do not have null addresses or values outside of boundary condition